packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -519,6 +547,45 @@ mixin WidgetsBinding on BindingBase, ServicesBinding, SchedulerBinding, GestureB
     return Future<void>.value();
   }
 
+  final Map<PreHotRestartCallback, String> _hotRestartCallbacks = <PreHotRestartCallback, String>{};
+
+  /// Register a callback that will be invoked before a hot restart is called.
+  ///
+  /// In non-debug modes this method is a no-op. This can be used to release native
+  /// resources acquired through platform channels or `dart:ffi`. Future returning
+  /// callbacks will be awaited, allowing for async tear downs.
+  ///
+  /// The following sample code shows how to use registerHotRestartCallback to handle
+  /// tearing down a native resource acquired through `dart:ffi`. In this example, if
+  /// the `context` pointer is not passed through to the `_destroyContext` function before
+  /// a hot restart, the application will crash after a hot restart.
+  ///
+  /// ```dart
+  /// // ignore_for_file: always_specify_types
+  /// import 'dart:ffi';
+  /// import 'package:flutter/widgets.dart';
+  ///
+  /// final DynamicLibrary _lib = DynamicLibrary.open('some_native_lib.dll');
+  /// final Pointer Function() _createContext = _lib.lookupFunction<Pointer Function(), Pointer Function()>('Native_create');
+  /// final void Function(Pointer) _destroyContext = _lib.lookupFunction<Void Function(Pointer), void Function(Pointer)>('Native_destroy');
+  ///
+  /// class NativeResourceService {
+  ///   NativeResourceService() {
+  ///     WidgetsBinding.instance.registerHotRestartCallback(() => _destroyContext(_context), debugLabel: 'NativeResourceService');
+  ///   }
+  ///
+  ///   /// Acquire native resources that must be released before they can
+  ///   /// be re-acquired.
+  ///   late final Pointer _context = _createContext();
+  /// }
+  /// ```
+  void registerHotRestartCallback(PreHotRestartCallback callback, {String debugLabel = 'unknown'}) {
```

Comment:
this name should start with "debug" to indicate that it only does something in debug mode.